### PR TITLE
Filter public link share data

### PR DESCRIPTION
**Pull request by settermjd**
## Description

Filter data passed to the script that emails a user about a public link share.

## Related Issue

https://github.com/owncloud/security-tracker/issues/255

## Motivation and Context

It closes an HTML injection vulnerability.

## How Has This Been Tested?

There is a limited ability to suitably test `/ocs/ajax/share.php`. Given that, I created tests for the class that I developed. Then, I manually tested the changes via the browser.

## Screenshots (if appropriate):

Relevant screenshots are contained in the related issue.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by tomneedham** [comment]
We should validate the email addresses with https://github.com/owncloud/core/blob/master/lib/private/Mail/Mailer.php#L118

**Comment by settermjd** [comment]
I deliberately avoided adding validation, as I'm not that familiar with the rest of the codebase, and felt that that may lead down a dark rabbit hole. However, I'm more than happy to add that in a separate PR.

**Comment by settermjd** [comment]
I made minor amendments. Any input @tomneedham @PVince81 @DeepDiver1975 would be appreciated.

File: composer.json [hunk 1]
```diff
@@ -58,6 +58,8 @@
         "symfony/polyfill-php70": "^1.0",
         "lukasreschke/id3parser": "^0.0.1",
         "sabre/dav": "^3.2",
-        "deepdiver/zipstreamer": "^1.1"
+        "deepdiver/zipstreamer": "^1.1",
+        "zendframework/zend-inputfilter": "^2.8",
```
Comment (PVince81) [inline composer.json]: @DeepDiver1975 are we ok to bring in these new libs ?

File: lib/private/Share/Filters/MailNotificationFilter.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace OC\Share\Filters;
+
+use Zend\InputFilter\Factory as InputFilterFactory;
+
+class MailNotificationFilter {
```
Comment (PVince81) [inline lib/private/Share/Filters/MailNotificationFilter.php]: PHPDoc, what is this class about ?

Comment (settermjd) [reply lib/private/Share/Filters/MailNotificationFilter.php]: Will get that updated.

**Comment by settermjd** [comment]
@PVince81, file and class dockblock comments added. Thanks for spotting that.

File: composer.json (hunk 1 above)
Comment (DeepDiver1975) [reply composer.json]: license is  BSD-3-Clause - need to chat with @hodyroff about this

**Comment by DeepDiver1975** [comment]
@ownclouders rebase

**Comment by ownclouders** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by ownclouders** [comment]
Automated rebase failed! Please rebase your pull request manually via the command line.

Reason:
```
Command: git rebase base/master

Exit Code: 128

Cause:

error: Failed to merge in the changes.


```

File: composer.json (hunk 1 above)
Comment (hodyroff) [reply composer.json]: License is good, need to add to our list.

**Comment by PVince81** [comment]
@settermjd can you rebase ? Seems the bot is not able today 😄 (job security for professional rebasers FTW)

**Review by DeepDiver1975 — APPROVED** [review]

**Comment by settermjd** [comment]
@PVince81 can do.

**Comment by settermjd** [comment]
@PVince81, merged with master.

File: composer.lock [hunk 3]
```diff
@@ -2246,16 +2326,16 @@
         },
         {
             "name": "sabre/vobject",
-            "version": "4.1.3",
+            "version": "4.1.4",
```
Comment (DeepDiver1975) [inline composer.lock]: this should not have happened

File: composer.lock [hunk 4]
```diff
@@ -3269,12 +3644,12 @@
             "source": {
                 "type": "git",
                 "url": "https://github.com/minkphp/MinkSelenium2Driver.git",
-                "reference": "46404b04c761feff4feab4d8d75957225e21a73c"
+                "reference": "93474c65a2a7bf959200ab5f7a14cc450645c185"
```
Comment (DeepDiver1975) [inline composer.lock]: same - please do not run composer update

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

File: composer.lock (hunk 4 above)
Comment (settermjd) [reply composer.lock]: Fair call. Will change.

**Comment by settermjd** [comment]
@DeepDiver1975; file revised.

**Review by DeepDiver1975 — APPROVED** [review]

**Comment by DeepDiver1975** [comment]
I assume we want to backport this ... @settermjd

**Comment by DeepDiver1975** [comment]
... will take care

**Comment by DeepDiver1975** [comment]
backport open

**Pull request merged**